domain/model/service/service.go

```diff
@@ -709,3 +730,69 @@ func (s WatchableService) WatchModel(ctx context.Context, modelUUID coremodel.UU
 		}),
 	)
 }
+
+// WatchModelCloudCredential returns a new NotifyWatcher watching for changes that
+// result in the cloud spec for a model changing. The changes watched for are:
+// - updates to model cloud.
+// - updates to model credential.
+// - changes to the credential set on a model.
+// The following errors can be expected:
+// - [modelerrors.NotFound] when the model is not found.
+func (s *WatchableService) WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (watcher.NotifyWatcher, error) {
+	if err := modelUUID.Validate(); err != nil {
+		return nil, errors.Errorf("invalid model UUID watching model cloud credential: %w", err)
+	}
+
+	// Get the model's cloud and credential UUID to use in the watcher filters.
+	cloudUUID, credentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+	if err != nil {
+		return nil, errors.Errorf("getting model cloud and credential: %w", err)
+	}
+
+	var lastSeenCredentialUUID atomic.Pointer[credential.UUID]
+	lastSeenCredentialUUID.Store(&credentialUUID)
+
+	// The mapper is used to filter out any model events where the credential UUID has not changed.
+	mapper := func(ctx context.Context, txnRunner database.TxnRunner, events []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		// Get the current model credential UUID.
+		_, currentModelCredentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+
+		var out []changestream.ChangeEvent
+		for _, e := range events {
+			wantEvent := false
+			switch e.Namespace() {
+			case "cloud":
+				wantEvent = true
+			case "model":
+				if *lastSeenCredentialUUID.Load() != currentModelCredentialUUID {
+					wantEvent = true
+					lastSeenCredentialUUID.Store(&currentModelCredentialUUID)
+				}
+			case "cloud_credential":
+				wantEvent = currentModelCredentialUUID.String() == e.Changed()
+			}
+			if wantEvent {
+				out = append(out, e)
+			}
+		}
+		return out, nil
+	}
+
+	result, err := s.watcherFactory.NewNotifyMapperWatcher(
+		mapper,
+		eventsource.PredicateFilter("model", changestream.Changed, func(v string) bool {
+			return v == modelUUID.String()
```

Comment:
These can now be dried up a little with the `EqualsPredicate` added by @jack-w-shaw.